### ISSUE TITLE
Custom line-height line-spacing results in cutoff

### ISSUE THREAD

**Issue by samimb** [minad/corfu#329]
There is an issue with Corfu and specifically with running a custom `line-height` and `line-spacing`
As seen in the screenshot below the line spacing and height are set `(setq default-text-properties '(line-spacing 0.35 line-height 1.25))` and Corfu cuts the results.

<img width="612" alt="Screenshot 2023-05-13 at 00 10 33" src="https://github.com/minad/corfu/assets/4834107/2a1b02b0-67a6-4c15-a77f-c506fbb77a50">
￼
When not setting it it behaves as intended

<img width="612" alt="Screenshot 2023-05-14 at 19 12 41" src="https://github.com/minad/corfu/assets/4834107/39f8a5a3-7b24-4fad-b546-a46232fd9a28">

As seen here, company mode works just as expected with a custom line height and spacing.
￼
<img width="612" alt="Screenshot 2023-05-13 at 00 11 38" src="https://github.com/minad/corfu/assets/4834107/a3d15c83-38f9-4c33-91ce-3cee862aa3ed">

The config for corfu is the following. 

```
(require 'corfu)
(require 'kind-icon)

(setq corfu-cycle t)
(setq corfu-auto t)
(setq corfu-auto-prefix 1)
(setq corfu-auto-delay 0.0)
(setq corfu-echo-delay 0.0)
(setq corfu-popupinfo-delay 0.0)
(setq corfu-count 10)
(setq corfu-scroll-margin 5)
(setq corfu-echo-documentation nil)

(defun my/enable-corfu ()
  "Enable Corfu mode on prog-mode hooks"
  (corfu-mode 1))
(add-hook 'prog-mode-hook #'my/enable-corfu)

(add-hook 'corfu-mode-hook #'corfu-popupinfo-mode)

(setq tab-always-indent 'complete)

(setq kind-icon-use-icons t)
(setq kind-icon-default-face 'corfu-default)
(setq kind-icon-blend-background nil)
(setq kind-icon-blend-frac 0.08)
(add-to-list 'corfu-margin-formatters #'kind-icon-margin-formatter)
```

**Comment by minad** [comment]
The variable `default-text-properties` should better not be used. I've never seen anyone adjusting it. You can use this instead:

~~~ elisp
(setq line-spacing 0.35)
~~~

**Comment by samimb** [comment]
> The variable `default-text-properties` should better not be used. I've never seen anyone adjusting it. You can use this instead:
> 
> ```emacs-lisp
> (setq line-spacing 0.35)
> ```

`default-text-properties` allows me to set both height and spacing.
Both of which centres the "spacing". `line-height` is not enough to set as it only sets the spacing from the bottom where as hight does so from the top. 

```
(setq line-spacing 0.35)
(setq line-height 1.25)
```
only sets one of them leaving the other unset for some reason, hence why `default-text-properties` is used here.
https://www.gnu.org/software/emacs/manual/html_node/elisp/Examining-Properties.html#index-default_002dtext_002dproperties

**Comment by minad** [comment]
Yes, I see. It is unfortunate that `line-height` is not a variable. It seems that taking `default-text-properties` into account would require some changes to the font height computation in `corfu--popup-show`. 

If you use this code from https://github.com/minad/corfu/issues/149#issuecomment-1179383404, the `line-height` will only be adjusted in the main buffer and not in the Corfu buffer.

~~~ elisp
(defun set-bigger-spacing ()
  (setq-local default-text-properties '(line-spacing 0.25 line-height 1.25)))
(add-hook 'text-mode-hook 'set-bigger-spacing)
(add-hook 'prog-mode-hook 'set-bigger-spacing)
~~~

**Comment by minad** [comment]
I tried to use `line-pixel-height` instead of `default-line-height` but this still somehow messes up the positioning. 

I tried a global setting `(setq default-text-properties '(line-spacing 0.25 line-height 1.25))` in my setup and this messes up a lot of other things. The issue is that `line-height` is only taken into account if a line ends with a final newline. This leads to an odd effect for the last line in a buffer. Last lines in buffers are commonly edited for example in shells, scratch and the minibuffer.

My conclusion is that using `default-text-properties` to adjust the line height is not a generally supported and recommended setting. In the Spacemacs thread linked above someone mentioned a patch for Emacs. This should probably be resubmitted to emacs-devel or the Emacs bug tracker.

Until then, I treat this issue as wontfix in Corfu.

**Comment by samimb** [comment]
Thank you for digging into this either way! Much appreciated!

On my phone at the moment so cannot see the actual patch that was submitted. 
Will see if I can put something on the bug tracker as Emacs does have some shortcomings with handling fonts properly.

**Comment by samimb** [comment]
> I tried to use `line-pixel-height` instead of `default-line-height` but this still somehow messes up the positioning.
> 
> I also experimented with a global setting `(setq default-text-properties '(line-spacing 0.25 line-height 1.25))` in my setup and this messes up a lot of other things. The issue is that `line-height` is only taken into account if a line ends with a final `\n` character. This leads to odd effects for the last line in a buffer. Last lines in buffers are commonly edited for example in shells, scratch and the minibuffer.
> 
> My conclusion is that using `default-text-properties` to adjust the line height is not a generally supported and recommended setting. In the Spacemacs thread linked above someone mentioned a patch for Emacs. This should probably be resubmitted to emacs-devel or the Emacs bug tracker, in order to support text centering properly.
> 
> Until then, I treat this issue as wontfix in Corfu.

I've created a ***very*** simple patch that seems to work for 30.0.50, current HEAD.
I wanted to check first though. What kind of strange behaviour more exactly did you encounter with `line-height`, if you don't mind? 
Just want to cover grounds before I try to submit the patch for review.

**Comment by minad** [comment]
Thanks, it would be nice to have text centering in Emacs via an additional `line-spacing-above` or `line-height` variable. I don't think I can be of much help. Maybe ask on the Spacemacs thread or send a mail to the people involved on the old emacs-devel thread? I also suggest that you send your patch to the Emacs bug tracker and not to the development mailing list. This way it is more likely that the improvement will actually land in Emacs.

**Comment by samimb** [comment]
Hey Minad,

My C skills are obviously lacking here :). Cause it is not 100% working (not asking for your help really, perhaps some potential insight on why the patch is still breaking on Corfu)

I wanted to have a look at the `line-spacing` variable as to me, it is slightly misleading. Spacing should apply both on top and bottom of the line, otherwise it would arguably be `line-spacing-bottom`. 

My patch is very simple in terms of just "fixing" the `line-spacing` to add top "padding" as well.

```
diff --git a/src/buffer.c b/src/buffer.c
index 7951b7ac322..0cec770ed98 100644
--- a/src/buffer.c
+++ b/src/buffer.c
@@ -5852,7 +5852,7 @@ from (abs POSITION).  If POSITION is positive, point was at the front
   DEFVAR_PER_BUFFER ("line-spacing",
 		     &BVAR (current_buffer, extra_line_spacing), Qnumberp,
 		     doc: /* Additional space to put between lines when displaying a buffer.
-The space is measured in pixels, and put below lines on graphic displays,
+The space is measured in pixels, divided by 2 and put above and below lines on graphic displays,
 see `display-graphic-p'.
 If value is a floating point number, it specifies the spacing relative
 to the default frame line height.  A value of nil means add no extra space.  */);
diff --git a/src/xdisp.c b/src/xdisp.c
index 982e1b3c103..c5d2934954f 100644
--- a/src/xdisp.c
+++ b/src/xdisp.c
@@ -23306,7 +23306,9 @@ append_space_for_newline (struct it *it, bool default_face_p)
 		}
 	      if (extra_line_spacing > 0)
 		{
-		  it->descent += extra_line_spacing;
+		  int split_line_spacing = extra_line_spacing / 2;
+		  it->descent += split_line_spacing;
+		  it->ascent += split_line_spacing;
 		  if (extra_line_spacing > it->max_extra_line_spacing)
 		    it->max_extra_line_spacing = extra_line_spacing;
 		}
@@ -32672,7 +32674,9 @@ gui_produce_glyphs (struct it *it)
 
   if (extra_line_spacing > 0)
     {
-      it->descent += extra_line_spacing;
+      int split_line_spacing = extra_line_spacing / 2;
+      it->descent += split_line_spacing;
+      it->ascent += split_line_spacing;
       if (extra_line_spacing > it->max_extra_line_spacing)
 	it->max_extra_line_spacing = extra_line_spacing;
     }
```


Now, the same problem as with `line-height` appears, as with the first screenshot in this issue.
Though, I've mainly tested it with Corfu thus far.

**Comment by minad** [comment]
It is unlikely that such a change will get into Emacs because it is not backward compatible. You could introduce a new variable `line-spacing-above`. To make this consistent, add a new alias `line-spacing-below` for `line-spacing`. This way we will get both backward compatibility and consistency.

Another option would be to introduce another variable `line-spacing-position` which changes the meaning of `line-spacing`. The variable could take three values `center`, `top` and `bottom` (the backward compatible default).

In order for your patch to work with Corfu you must make sure that the line height computation also takes the new variable into account, see `default-line-height`. You should grep the Emacs code base and check every occurrence of `line-spacing`.

But again, this is not the right place to discuss this. Please send your patch to the Emacs bug tracker and ping the people involved in the old patch.

**Comment by samimb** [comment]
I ended up patching my fonts to add native line spacing until 31 is released. 

These patched fonts are only used in emacs at the moment.
